rollup/types/block.go

```diff
@@ -0,0 +1,192 @@
+package types
+
+import (
+	"encoding/binary"
+	"errors"
+	"math"
+	"math/big"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/log"
+)
+
+// CalldataNonZeroByteGas is the gas consumption per non zero byte in calldata.
+const CalldataNonZeroByteGas = 16
+const blockContextByteSize = 60
+
+// GetKeccak256Gas calculates the gas cost for computing the keccak256 hash of a given size.
+func GetKeccak256Gas(size uint64) uint64 {
+	return GetMemoryExpansionCost(size) + 30 + 6*((size+31)/32)
+}
+
+// GetMemoryExpansionCost calculates the cost of memory expansion for a given memoryByteSize.
+func GetMemoryExpansionCost(memoryByteSize uint64) uint64 {
+	memorySizeWord := (memoryByteSize + 31) / 32
+	memoryCost := (memorySizeWord*memorySizeWord)/512 + (3 * memorySizeWord)
+	return memoryCost
+}
+
+// WrappedBlock contains the block's Header, Transactions and WithdrawTrieRoot hash.
+type WrappedBlock struct {
+	Header               *types.Header         `json:"header"`
+	Transactions         []*types.Transaction  `json:"transactions"`
+	WithdrawRoot         common.Hash           `json:"withdraw_trie_root"`
+	RowConsumption       *types.RowConsumption `json:"row_consumption"`
+	txPayloadLengthCache map[common.Hash]uint64
+}
+
+// BlockContext represents the essential data of a block in the ScrollChain.
+// It provides an overview of block attributes including hash values, block numbers, gas details, and transaction counts.
+type BlockContext struct {
+	BlockHash       common.Hash
+	ParentHash      common.Hash
+	BlockNumber     uint64
+	Timestamp       uint64
+	BaseFee         *big.Int
+	GasLimit        uint64
+	NumTransactions uint16
+	NumL1Messages   uint16
+}
+
+// Encode encodes the WrappedBlock into RollupV2 BlockContext Encoding.
+func (w *WrappedBlock) Encode(totalL1MessagePoppedBefore uint64) ([]byte, error) {
+	bytes := make([]byte, blockContextByteSize)
+
+	if !w.Header.Number.IsUint64() {
+		return nil, errors.New("block number is not uint64")
+	}
+
+	numL1Messages := w.NumL1Messages(totalL1MessagePoppedBefore)
+	if numL1Messages > math.MaxUint16 {
+		return nil, errors.New("number of L1 messages exceeds max uint16")
+	}
+
+	numL2Transactions := w.NumL2Transactions()
+	numTransactions := numL1Messages + numL2Transactions
+	if numTransactions > math.MaxUint16 {
+		return nil, errors.New("number of transactions exceeds max uint16")
+	}
+
+	binary.BigEndian.PutUint64(bytes[0:], w.Header.Number.Uint64())
+	binary.BigEndian.PutUint64(bytes[8:], w.Header.Time)
+	if w.Header.BaseFee != nil {
```

Comment:
Is this the only place to be updated to support EIP-1559 in block context encoding? @zimpha 
previously it was a comment here:
```
// TODO: [16:47] Currently, baseFee is 0, because we disable EIP-1559.
```